plugins/filter_aws/aws.c

```diff
@@ -370,6 +400,208 @@ static int get_vpc_metadata(struct flb_filter_aws *ctx)
     return ret;
 }
 
+void flb_filter_aws_tags_destroy(struct flb_filter_aws *ctx)
+{
+    size_t i;
+    if (!ctx) {
+        return;
+    }
+    if (ctx->tag_keys) {
+        for (i = 0; i < ctx->tags_count; i++) {
+            if (ctx->tag_keys[i]) {
+                flb_sds_destroy(ctx->tag_keys[i]);
+            }
+        }
+        flb_free(ctx->tag_keys);
+        ctx->tag_keys = NULL;
+    }
+    if (ctx->tag_values) {
+        for (i = 0; i < ctx->tags_count; i++) {
+            if (ctx->tag_values[i]) {
+                flb_sds_destroy(ctx->tag_values[i]);
+            }
+        }
+        flb_free(ctx->tag_values);
+        ctx->tag_values = NULL;
+    }
+    if (ctx->tag_keys_len) {
+        flb_free(ctx->tag_keys_len);
+    }
+    ctx->tag_keys_len = NULL;
+    if (ctx->tag_values_len) {
+        flb_free(ctx->tag_values_len);
+    }
+    ctx->tag_values_len = NULL;
+    ctx->tags_count = 0;
+}
+
+/* Get EC2 instance tag keys from /latest/meta-data/tags/instance.
+ * Initializes ctx->tags_count, ctx->tag_keys and ctx->tag_keys_len.
+ */
+static int get_ec2_tag_keys(struct flb_filter_aws *ctx)
+{
+    int ret;
+    flb_sds_t tags_list = NULL;
+    size_t len = 0;
+    size_t tag_index = 0;
+    size_t tag_start = 0;
+    size_t tag_end = 0;
+    flb_sds_t tag_key;
+    size_t tag_key_len;
+    size_t i;
+
+    /* get a list of tag keys from the meta data server */
+    ret = get_metadata(ctx, FLB_FILTER_AWS_IMDS_INSTANCE_TAG, &tags_list, &len);
+    if (ret < 0) {
+        if (ret == -2) { /* if there are no tags, response status code is 404 */
+            flb_plg_warn(ctx->ins, "tags not available in the EC2 instance metadata");
```

Comment:
So this code means there are no tags on the instance right? Is that guaranteed or just implied?

If so/depending, can we make the error message a little more user friendly like: "EC2 instance metadata tag request returned 404. This likely indicates your instance has no tags or the EC2 tagging metadata API is not enabled".

Would something like that be more accurrate?

Reply:
> So this code means there are no tags on the instance right? Is that guaranteed or just implied?

I don't know if this is guaranteed. I've performed a few tests on the EC2 and in both cases:
  1) exposing tags disabled in the metadata,
  2) exposing tags enabled in the metadata but no tags added,
this endpoint returned 404.

For this reason, I assume it won't be changed due the scale of AWS (Hyrum's law applied).

> Would something like that be more accurrate?

Yes, I've added the proposed warning message. https://github.com/fluent/fluent-bit/pull/6065/commits/43ad93b368c8ad28526fbdf5041b046e5dff993d